New error when two different ICs are specified

DanielYankura opened the pull request:
-Error thrown if an initial_condition parameter is provided along with an initial_from_file_var parameter for the same variable.

closes #20944

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
This added error check will make debugging input files easier.

## Design
<!--A concise description (design) of the enhancement.-->
Previously an error was only thrown if two initial_condition parameters were supplied for the same variable. Now an error is also thrown if an initial_from_file_var is also provided.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

DanielYankura committed:
New error when two different ICs are specified

-Error thrown if an initial_condition parameter is provided along with an initial_from_file_var parameter for the same variable.

closes #20944

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2631908/), step Docs: sync website on 0db280d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29672/site)

This comment will be updated on new commits.

GiudGiud commented on framework/src/actions/AddVariableAction.C:
```diff
@@ -132,6 +133,12 @@ AddVariableAction::init()
 
+  if (_pars.isParamSetByUser("initial_condition") &&
+      _pars.isParamSetByUser("initial_from_file_var"))
+    mooseError("Two initial conditions have been provided for the variable ",
+               name(),
+               ". One from the user and one from the restart input file.");
```
```suggestion
    paramError("initial_condition", "Two initial conditions have been provided for the variable ",
               name(),
               " using the 'initial_condition' and 'initial_from_file_var' parameters. Please remove one of them.");
```

GiudGiud reviewed (changes requested):
Good idea!

DanielYankura committed:
Made changes to error message

-error message is worded more clearly now
-initial_from_file_var is now a transfer param
-test was updated to reflect new error message

GiudGiud commented:
let's squash formatting commits with the commit introducing the formatting problem

or even better, install the format hooks (see scripts/) to avoid failing the precheck

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2637795/), step Generate coverage on f61fac0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2df091</th>
      <th colspan=4>#29672 f61fac</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.25%</td>
      <td>85.25%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108039</td>
      <td>108043</td>
      <td>+4</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18699</td>
      <td>18700</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29672/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29672/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>2df091</th>
      <th colspan=4>#29672 f61fac</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.14%</td>
      <td>85.13%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7895</td>
      <td>7894</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1378</td>
      <td>1379</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29672/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29672/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29672/coverage/xfem)

</details>



This comment will be updated on new commits.

DanielYankura committed:
added detail to test case

GiudGiud commented on test/tests/restart/restart_diffusion/gold/exodus_refined_restart_1.e:
let's not regold those files. 

you can either:
- keep your new test design but use Outputs/hide='v' to avoid needing to regold
- return back to the previous test and simply add the variable v in the command line to the test that is checking for restart + ics

DanielYankura replied:
So I tried both of those methods and in both cases the simulation ran instead of erroring out. I'm wondering if adding the variable later by command line, or by hiding it in previous simulations, it doesn't get recognized as being a restarted variable and so it never errors?

DanielYankura replied:
What if instead I change my error into a warning saying that one of the input parameters will be overwritten. Then for this test case I can use the original input files and get the same error?

GiudGiud replied:
warnings get ignored. the error here is better

> So I tried both of those methods and in both cases the simulation ran instead of erroring out

well this is something we need to investigate. The errors, both the existing one and the one you are adding, need to be reliably hit

GiudGiud replied:
> I'm wondering if adding the variable later by command line, or by hiding it in previous simulations, it doesn't get recognized as being a restarted variable and so it never errors?

It needs to be in the exodus file to count as "restarted" so it needs to be part of the first simulation. 
However for your case, you can set a second variable from the command line with something like:
```
Variables/v/initial_condition=3 Variables/v/initial_from_file_var='u'
```

DanielYankura replied:
So I did some digging and I think I know why those two methods weren't working. So the error check in `FEProblemBase.C' only works for initial conditions applied to restarted variables so adding in a variable through the command line that isn't being read in from an output file won't make it error. I did try moving my test over from `AddVariableAction.C` to `FEProblemBase.C` but I haven't been able to get it to properly check for an initial condition. If I can get that to work though then I can have my test get hit after the other.

GiudGiud replied:
ok we did not need to set another `initial_from_file_var` in the end, it was more about creating the IC differently. 

I pushed an alternative commit. should work

GiudGiud committed:
Use a regular IC to test IC+restart without triggering the initial_condition + initial_from_file_var check

GiudGiud commented:
Failures are two unrelated timeouts. @DanielYankura if you're ok with the change I ll merge

DanielYankura commented:
Sounds good. I took a look over the changes you made and I am good with them if you want to merge!
